File: ASSI/prc.py
class N:
    def __init__(self, d):
        self.d = d
        self.n = None

class ll:
    def __init__(self):
        self.h = None

    def add(self, d):
        # Directly add at the end by finding the last node
        if self.h is None:
            self.h = N(d)
        else:
            # Move `self.h` to end but reset it afterward
            original_h = self.h
            while self.h.n is not None:
                self.h = self.h.n
            self.h.n = N(d)
            self.h = original_h

    def dis(self):
        # Display without changing `self.h`
        if self.h is None:
            print('e')
        else:
            original_h = self.h
            while self.h is not None:
                print(self.h.d, end='->')
                self.h = self.h.n
            print('n')
            self.h = original_h

    def se(self, v):
        # Search without changing `self.h`
        p = 0
        found = False
        original_h = self.h
        while self.h is not None:
            if self.h.d == v:
                print(f'{v} at {p}')
                found = True
                break
            self.h = self.h.n
            p += 1
        if not found:
            print('!found')
        self.h = original_h

    def rem(self, v):
        # Remove node without changing `self.h`
        if self.h is None:
            print('e')
        else:
            if self.h.d == v:
                # Special case: remove head node
                self.h = self.h.n
                print('r')
            else:
                # Find node to remove
                original_h = self.h
                prev = None
                while self.h is not None:
                    if self.h.d == v:
                        prev.n = self.h.n
                        print('r')
                        break
                    prev = self.h
                    self.h = self.h.n
                else:
                    print('!find')
                self.h = original_h

l = ll()
l.add(1)
l.add(2)
l.se(3)
l.se(1)
l.rem(1)
l.se(1)
l.dis()

File: ASSI/prc.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data){
        this-> data = data; // . *  to ->
        this-> next = nullptr; // this.next* == this->next
    }
};

class LinkedList {
private:
    Node* head;
public:
    LinkedList(){
        head = nullptr;
    }

    void add(int data){
        Node*  n = new Node(data);
        if (head  == nullptr){
            head = n;
        }
        else {
            Node* t = head;
            while ( t-> next != nullptr){
                t = t->next;
            }
            t->next = n;
        }
    }

    void view(){
        if (head == nullptr){
            cout << "empty" << endl;
        }
        else {
            Node* t = head;
            while ( t != nullptr){
                cout << t->data << " -> ";
                t = t->next;
            }
            cout << "None" << endl;
        }
    }

    void search(int v){
        int pos = 0;
        Node* t = head;
        while ( t != nullptr){
            if (t-> data == v){
                cout << v << " found at " << pos << endl;
                return;
            }
            t = t -> next;
            pos++;
        }
        cout << "not found" << endl; 
    }

    void remove ( int v){
        if (head == nullptr){
            cout << "empty" << endl;
            return;
        }
        if (head -> data == v){
            Node* d = head;
            head = head ->next;
            delete d;
            cout << "removed" << endl;
            return;
        }

        Node* t = head;
        while ( t -> next != nullptr){
            if ( t -> next -> data == v ){
                Node* d = t-> next;
                delete d;
                cout << "removed" << endl;
                return;
            }
            t = t->next;
        }
        cout << "not found" << endl;
    }
};

int main(){
    LinkedList l;
    l.add(1);
    l.add(2);
    l.view();
    l.search(2);
    l.add(3);
    l.view();
    l.search(1);

    return 0;
}

